### PR TITLE
implemented native SASL/JAAS support for plaintext authentication

### LINKED ISSUE #8168: Kafka Client does not support SASL with JASS
**Describe the bug**

Kafka SASL is not supported in native mode.

Even with the added set of classes, it fails with:

```
Caused by: javax.security.sasl.SaslException: Cannot get userid/password [Caused by javax.security.auth.callback.UnsupportedCallbackException: Could not login: the client is being asked for a password, but the Kafka client code does not currently support obtaining a password from the user.]
 at com.sun.security.sasl.ClientFactoryImpl.getUserInfo(ClientFactoryImpl.java:157)
        at com.sun.security.sasl.ClientFactoryImpl.createSaslClient(ClientFactoryImpl.java:94)
        at javax.security.sasl.Sasl.createSaslClient(Sasl.java:433)
        at org.apache.kafka.common.security.authenticator.SaslClientAuthenticator.lambda$createSaslClient$0(SaslClientAuthenticator.java:183)
        at java.security.AccessController.doPrivileged(AccessController.java:144)
        at javax.security.auth.Subject.doAs(Subject.java:423)
```

**Current Status**

I have added the following code to `io.quarkus.kafka.client.deployment.KafkaProcessor`.

```
reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, SaslClientCallbackHandler.class.getName()));
        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, DefaultLogin.class.getName()));
        reflectiveClass
                .produce(new ReflectiveClassBuildItem(true, false,
                        AbstractLogin.DefaultLoginCallbackHandler.class.getName()));
        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, PlainLoginModule.class.getName()));
```

### PR DESCRIPTION
**Pull request by schulzp**
fixes #8168

### CODE REVIEW

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (cescoffier) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:81]: This list was about serde, LoginModule might go in another list (it's actually not build-in as it comes from `javax.security.auth.spi.LoginModule`)

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (cescoffier) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:125]: I would add the LoginModule directly here.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java
Comment (cescoffier) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java:28]: How is this related to the issue? I may miss a link.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
Comment (cescoffier) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:19]: If I understand correctly you copy the "kafka" configuration and configure the health check admin client with them.
Is it really related to this PR?

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
Comment (cescoffier) [inline extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:31]: It can be confusing as the Kafka Client is not produced by default. Typically, it would not cover the configuration coming from reactive messaging. Existing applications create their properties directly (following the Kafka documentation).

File: integration-tests/kafka/pom.xml
Comment (cescoffier) [inline integration-tests/kafka/pom.xml:159]: Why is this passed here and not in the Map produced by the Kafka Test Resource?

File: integration-tests/kafka/pom.xml
Comment (cescoffier) [inline integration-tests/kafka/pom.xml:110]: Why are these properties configured here and not in the Kafka Test Resource?

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (cescoffier) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:153]: I believe we should be able to avoid this new bean. 
It seems to be convenient for the tests, but the Kafka Test Resource can configure properties (that's the Map returned by the `start` method)

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java
Comment (schulzp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java:28]: the previous `@PostConstruct` ran into NPEs for the `@Inject`ed `KafkaRuntimeConfig`. Using constructor injection it worked.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (schulzp) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:125]: Actually, I need all the sub-classes, for example (but not limited to):
* PlainLoginModule (org.apache.kafka.common.security.plain)

The list of sub-classes also contains:
* ExternalCertificateLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* Krb5LoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* PropertiesLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* DigestLoginModule (org.apache.zookeeper.server.auth)
* LDAPLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* ScramLoginModule (org.apache.kafka.common.security.scram)
* NTLoginModule (com.sun.security.auth.module) 
* GuestLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* InVMLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* FileLoginModule (com.sun.jmx.remote.security)
* JndiLoginModule (com.sun.security.auth.module)
* AbstractKeycloakLoginModule (org.keycloak.adapters.jaas)
* LdapLoginModule (com.sun.security.auth.module)
* OAuthBearerLoginModule (org.apache.kafka.common.security.oauthbearer)
* UnixLoginModule (com.sun.security.auth.module)
* Krb5LoginModule (com.sun.security.auth.module)
* CertificateLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* KeyStoreLoginModule (com.sun.security.auth.module)

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
Comment (schulzp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:19]: I'd say so. The admin client needs the same security configuration as all consumers/producers. Otherwise authentication will fail once the broker is set up to expect it.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
Comment (schulzp) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:31]: Okay, in that case I could remove the KafkaRuntimeConfiguration again and would only have to make sure all the clients in integration-tests/kafa get the right set of properties. I just found it to be very cumbersome to inject them one by one, especially when extending the test case to more than plain authentication.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (schulzp) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:81]: Would a separate array/list suffice?

File: integration-tests/kafka/pom.xml
Comment (schulzp) [reply integration-tests/kafka/pom.xml:159]: This *instructs* the test resource to set up the broker to request authentication and to pass the required kafka-client properties on to quarkus.

File: integration-tests/kafka/pom.xml
Comment (schulzp) [reply integration-tests/kafka/pom.xml:110]: This instructs the test resource to set up the broker to request authentication and to pass the required kafka-client properties on to quarkus.

**Comment by schulzp** [comment]
@cescoffier, thank you for your review! I fixed a few things:
1. I followed your SASL/SSL example regarding the integration tests. You already named your tests SASL[...] despite only enabling SSL so I allowed myself to rename them into SSL[...] and used SASL for actual SASL/JAAS authentication.
2. I moved all native-SASL/JAAS related code into a separate `@BuildStep`.
3. I still don't know how to make `KafkaHealthCheck` or more specific its `AdminClient` aware of the properties like required to connect to a SASL/JAAS-enabled broker, for example, `sasl.jaas.config`, `sasl.mechanism`, and `security.protocol`.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (schulzp) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java]: Relocated native-SASL/JAAS code into separate build step.

File: integration-tests/kafka/pom.xml
Comment (schulzp) [reply integration-tests/kafka/pom.xml]: Implemented dedicated integration test, leaving the existing ones untouched.

File: integration-tests/kafka/pom.xml
Comment (schulzp) [reply integration-tests/kafka/pom.xml]: Implemented dedicated integration test, leaving the existing ones untouched.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java
Comment (cescoffier) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:165]: If we go with this, we would need to adapt the Reactive Messaging Kafka connector to use the same config. This would be a breaking change.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java
Comment (cescoffier) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java:28]: Why not using the static access to the configuration? That would remove the bean.

File: extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java
Comment (cescoffier) [reply extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:19]: Yes and no. You have applications using multiple brokers, and some parts of Quarkus use a different configuration format.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Almost there! Looks mostly good to me.

Still wondering about the new bean introduced to carry the config. I'm not against it, it just needs to be done carefully because the RM Kafka connector uses a different format and, because of project dependencies (Quarkus depends on the connector not the opposite), would not be able to depend on a Quarkus bean. My concern is that with such beans, the config would be quite different between the health checks and the connector. 

For AMQP we introduced the same kind of bean (but in the connector and with name support to handle the multi-broker case). 

Basically, what we would need is a bean producing a Map<String, ?>. If we use the `@Named` qualifier and set it to "default-kafka", we can imagine having the health check retrieving and using this bean and the RM connector retrieving and using this bean too. Would that work?

**Comment by schulzp** [comment]
Hi @cescoffier,

I've experimented a bit with code generation to cater for multiple (named) brokers but I keep running into alleged dead ends. Here is what I did:
* Create a `KafkaBrokerBuildTimeConfig` and add `defaultBroker` and `namedBrokers` to `KafkaBuildTimeConfig`
* Generate a `KafkaRuntimeConfigProducer extends AbstractKafkaRuntimeConfigProducer` with `@Produces @Named("kafka-config-{broker}") Map<String, ?> produce_{broker}() { ... }` per (named) broker.
* Generate a `KafkaHealthCheck_{broker}` with `@Inject @Named("kafka-config-{broker}") Map<String, ?> config`

Problems I ran into:
* Despite specifying a super class for the generated classes they cannot be cast to it (ClassCastException) when the `@Recorder` should call `AbstractKafkaRuntimeConfigProducer#setBuildTimeConfig`
* The `KafkaBrokerBuildTimeConfig` currently has only one `@ConfigProperty(name = "security.protocol")` to determine at build-time if I need to load all the SASL/JAAS-related classes. In a scenario where no broker uses a non-default `security.protocol` the list of named broker configurations is empty at build-time and to config/health-check gets generated. When I later specify `kafka.broker_a.boostrap.servers=...` and `kakfa.broker_b.boostrap.servers` at runtime that fails because there is no `KafkaRuntimeConfigProducer` which picks up the config for `broker_a` or `broker_b`.
* I only want to generate a health-check if the health feature is enabled. If I pass `ClassCreator#getClassName()` to `HealthBuildItem` it fails at runtime since the class-loader cannot find the generated class.

Is this the right approach? Any help would be appreciated.

**Comment by cescoffier** [comment]
Did you push the code? Maybe we should start with just a Map exposed using `@Named("default-kafka-broker")` and we revisit later. Note that this map should be exposed regardless of health as the RM connector would consume it.

**Comment by schulzp** [comment]
@cescoffier, pushed it now with runtime config being exposed as `@Named("default-kafka-broker") Map<String, Object>`. Hopefully that works for you.

**Comment by cescoffier** [comment]
Yes, that will work for me! Thanks! I will change the Kafka connector to consume this map.

**Comment by schulzp** [comment]
Hi @cescoffier, the CI pipelines passed. Do you still require any changes in the PR?

File: integration-tests/kafka/pom.xml
Comment (cescoffier) [inline integration-tests/kafka/pom.xml:200]: Any reason for this?

**Review by cescoffier — CHANGES_REQUESTED** [review]
One small comment.

Could you also rebase on latest master?

File: integration-tests/kafka/pom.xml
Comment (schulzp) [reply integration-tests/kafka/pom.xml]: Good catch, removed it.

**Comment by schulzp** [comment]
@cescoffier, done.

**Review by cescoffier — APPROVED** [review]

**Comment by schulzp** [comment]
Hi @cescoffier, rebased again, resolved conflicting change 0dc12512fa6b26aefd0fd85461a40a7377e74034 in `KafkaSASLTestResource`/`KafkaSSLTestResource`

**Comment by cescoffier** [comment]
@schulzp Waiting for CI, will merge once the CI says `ok`.

**Comment by cescoffier** [comment]
Thanks @schulzp !

**Pull request merged**
